### PR TITLE
Add bitwise distributed reduction ops

### PR DESCRIPTION
**Pull request by pietern**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26824 Add bitwise distributed reduction ops**

These ops are named after the bitwise reduction ops in MPI.

This is based on the work done by @knottb in #22449.

Closes #22449.

Differential Revision: [D17600210](https://our.internmc.facebook.com/intern/diff/D17600210)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 1]
```diff
@@ -113,7 +115,80 @@ class GlooStore : public ::gloo::rendezvous::Store {
...
+
+// Bitwise AND with SFINAE guard for integral types.
+template <
+    typename T,
+    typename std::enable_if<std::is_integral<T>::value, int>::type = 0>
+void band(void* c, const void* a, const void* b, size_t n) {
+  auto tc = static_cast<T*>(c);
+  auto ta = static_cast<const T*>(a);
+  auto tb = static_cast<const T*>(b);
+  for (size_t i = 0; i < n; i++) {
```
Comment (rohan-varma) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Do we need to validate that we won't get indexing errors here, or does Gloo's reduce function provide us with the correct `n`?

**Review by rohan-varma — APPROVED** [review]
Do we need to validate that we won't get indexing errors in the implementation of band, bor, bxor, or does Gloo's reduce function provide us with the correct `n`?

**Review by zhaojuanmao — APPROVED** [review]
lgtm, too

**Comment by pietern** [comment]
@rohan-varma This is guaranteed to be in range, as a subset of the total number of elements.

**Pull request closed without merging**